lua/lazyvim/plugins/extras/lang/java.lua

```diff
@@ -1,125 +1,143 @@
-                    vim.keymap.set("n", "<leader>tT", function() require("jdtls").pick_test({ bufnr = buffer }) end, { buffer = buffer, desc = "Run specific Test" })
-                    vim.keymap.set("n", "<leader>tt", function() require("jdtls").test_class({ bufnr = buffer }) end, { buffer = buffer, desc = "Run File" })
-                    vim.keymap.set("n", "<leader>tr", function() require("jdtls").test_nearest_method({ bufnr = buffer }) end, { buffer = buffer, desc = "Run nearest" })
-                  end
-                  if mason_registry.has_package("java-debug-adapter") then
-                    -- custom init for Java debugger
-                    require("jdtls").setup_dap({ hotcodereplace = "auto" })
-                    require("jdtls.dap").setup_dap_main_class_configs()
-                  end
-                  require("jdtls.setup").add_commands()
-                end),
-                cmd = cmd,
-                root_dir = root_dir,
-                init_options = {
-                  bundles = bundles,
-                }
-              }
-              local jdtls_opts = require("lazyvim.util").opts("nvim-jdtls")
-              require("jdtls").start_or_attach(vim.tbl_deep_extend("force", jdtls_opts or {}, jdtls_base_config))
-              require("which-key").register({ c = { x = { name = "Extract" } } }, { prefix = "<leader>" })
-            end,
-          })
-          return true -- avoid duplicate servers
+          return true
         end,
       },
     },
   },
+  {
+    "mfussenegger/nvim-jdtls",
+    dependencies = {
+      "neovim/nvim-lspconfig",
+      "mfussenegger/nvim-dap",
+      "hrsh7th/nvim-cmp",
+      "folke/which-key.nvim",
+    },
+    ft = { "java" },
+    config = function()
+      local project_name = vim.fn.fnamemodify(vim.fn.getcwd(), ":p:h:t")
+      local workspace_dir = vim.fn.stdpath("data") .. "/lsp/java/" .. project_name
+      local config = {
+        cmd = {
+          "java",
+          "-Declipse.application=org.eclipse.jdt.ls.core.id1",
+          "-Dosgi.bundles.defaultStartLevel=4",
+          "-Declipse.product=org.eclipse.jdt.ls.core.product",
+          "-Dlog.protocol=true",
+          "-Dlog.level=ALL",
+          "-Xmx1g",
+          "--add-modules=ALL-SYSTEM",
+          "--add-opens",
+          "java.base/java.util=ALL-UNNAMED",
+          "--add-opens",
+          "java.base/java.lang=ALL-UNNAMED",
+          "-jar",
+          vim.fn.expand("$MASON/share/jdtls/plugins/org.eclipse.equinox.launcher.jar"),
+          "-configuration",
+          vim.fn.expand("$MASON/share/jdtls/config"),
+          "-data",
+          workspace_dir,
+        },
+        root_dir = require("jdtls.setup").find_root({
+          ".git",
+          "mvnw",
+          "gradlew",
+          "pom.xml",
+          "build.gradle",
+          "build.gradle.kts",
+          "settings.gradle",
+          "settings.gradle.kts",
+          "build.xml",
+          ".project",
+        }),
```

Comment:
I just put every filetype in `require("lspconfig.server_configurations.jdtls").default_config` into the `root_dir`, 

The reason is the same as I explained above, exposing more detail can help user to understand what the config do, and help them to create their own config if they have more requirement

Add some more filetype that most user should have in the project root(such as `pom.xml`, `.git`)

Reply:
I'll just reiterate that finding the project root needs to happen before calculating the project name.

About the root detection method, `pom.xml` is already in the default list, see https://github.com/neovim/nvim-lspconfig/blob/3fe1e8de80b98c7a6b16f730711b5eafe84212e1/lua/lspconfig/server_configurations/jdtls.lua#L79-L89

For java projects specifically, `.git` is a false indicator because it can trip on submodules.

I think you should stick with calling the upstream-provided functions as this code did previously. It's counter-productive to reinvent things. If changes deserve to be made, consider proposing them to the lspconfig project.